Fixes #6502 - Add a list of ignored classes in the Puppet classes view

bastilian opened the pull request:
This PR addresses the issue reported in [#6502](http://projects.theforeman.org/issues/6502) by making ignored environments and classes visible when importing like so: 
![screen shot 2016-11-14 at 13 50 43](https://cloud.githubusercontent.com/assets/7757/20265689/cc00ff4c-aa72-11e6-8275-ddf8349a9d2f.png)

In case all available environments are ignored as full and no other changes are detected it will also mention these in the notice:
![screen shot 2016-11-14 at 14 02 51](https://cloud.githubusercontent.com/assets/7757/20265777/42721616-aa73-11e6-8288-0169a7b002a8.png)

domcleal commented on test/unit/puppet_class_importer_test.rb:
```diff
@@ -18,11 +18,67 @@ def setup
+      @proxy = smart_proxies(:puppetmaster)
+    end
+
+    context 'a spcefific environment is set' do
```
specific

domcleal commented on test/unit/puppet_class_importer_test.rb:
```diff
@@ -18,11 +18,67 @@ def setup
+      test "should contain only the specified environment in changes" do
+        importer = PuppetClassImporter.new(url: @proxy.url, env: 'foreman-testing')
+
+        assert importer.changes['new'].include?('foreman-testing')
```
Prefer assert_includes, refute_includes (better debug output on assertion failure).

domcleal commented on app/views/api/v1/import_puppetclasses/show.json.rabl:
```diff
@@ -25,3 +28,7 @@ end
   environment.name
 end
+
+node(:ignored_environment, :if => ->(environment) { @changed['ignored'][environment.name].present? && @changed['ignored'][environment.name].match(/_ignored_/) }) do |environment|
```
These changes should be made to API v2 - either only v2, or both v2 and v1 if you really wish to continue extending it.

domcleal commented on app/services/puppet_class_importer.rb:
```diff
@@ -94,6 +107,29 @@ def updated_classes_for(environment)
+  #  * ++class_name++:: {String} containing the class to be checked
+  #
+  def ignored_class?(class_name)
+    !ignored_classes.select do |filter|
```
Use `.any?` instead of `! .. select .. empty?`, it's faster and easier to read.

domcleal commented on app/controllers/concerns/foreman/controller/environments.rb:
```diff
@@ -21,7 +21,12 @@ def import_environments
+      notice_message = _("No changes to your environments detected")
+      if !@changed['ignored'].empty?
+        notice_message << "\n"
+        notice_message << _("Ignored Environments: %s") % @changed['ignored'].keys.join(', ')
```
1. Lower case "environments"
2. Use `.to_sentence` instead of `join(', ')`, it's locale-aware

domcleal commented on app/controllers/concerns/foreman/controller/environments.rb:
```diff
@@ -21,7 +21,12 @@ def import_environments
     else
-      notice _("No changes to your environments detected")
+      notice_message = _("No changes to your environments detected")
+      if !@changed['ignored'].empty?
```
.present? or unless

domcleal commented on test/controllers/api/v1/smart_proxies_controller_test.rb:
```diff
@@ -223,4 +223,18 @@ def setup_import_classes
     end
   end
+
+  test 'should contain ignored environments' do
```
ditto re API versions, should be in v2

domcleal commented on app/services/puppet_class_importer.rb:
```diff
@@ -94,6 +107,29 @@ def updated_classes_for(environment)
+  # Gives back the classes ignored for a given environment
+  #
+  # Params:
+  #  * ++environment++:: {String} name of the environment
```
Is this meant to be RDoc? (normally single `+` isn't it for tt?)

domcleal commented on app/services/puppet_class_importer.rb:
```diff
@@ -94,6 +107,29 @@ def updated_classes_for(environment)
+  #  * ++environment++:: {String} name of the environment
+  #
+  def ignored_classes_for(environment)
+    classes = proxy.classes(environment)
```
Can you use the existing query instead of making a second request to the smart proxy? The performance of this API in some configurations may be poor so it shouldn't be called unnecessarily.

It looks like the information is broadly available via `actual_classes`, but it will need a little refactoring to keep the ignored classes.

domcleal commented on app/services/puppet_class_importer.rb:
```diff
@@ -94,6 +107,29 @@ def updated_classes_for(environment)
+  #
+  def ignored_class?(class_name)
+    !ignored_classes.select do |filter|
+      filter =~ class_name
```
This duplicates the check in `actual_classes` but without the safety check for Regexp - it should be shared code.

domcleal commented:
@theforeman-bot ok to test

bastilian commented:
@domcleal thanks for reviewing! I've incorporated the changes you suggested.

There is an issue with moving the `ignored_environments` to `api/v2`.
`import_puppetclasses_common_controller.rb` only uses v1 templates see: [import_puppetclasses_common_controller.rb:37](https://github.com/theforeman/foreman/blob/acf4b6870c6d39fc2ea42811a713e4fe704be340/app/controllers/concerns/api/import_puppetclasses_common_controller.rb#L37)

Should I make the import_puppetclasses_common_controller.rb render the template matching the requested api version?

domcleal commented:
> Should I make the import_puppetclasses_common_controller.rb render the template matching the requested api version?

Yes, I think so.

bastilian commented:
I made it use `api_version` to use in the path for the templates. This should do the trick.

dLobatog commented:
Thanks @bastilian - overall :+1: on the implementation and tests, it worked pretty well. I found a small bug though:
- Set a 'truthy' or 'falsy' value as one of the ignored_environments, e.g: 

``` yml
:ignored:
  - no 
  - yes
```
- Try to import environments: you get a 500

The problem is that Foreman is reading those keys as actual Ruby yes or no values, instead of as strings, so that `@changed['ignored']` becomes `{"test"=>["_ignored_"], false=>["_ignored_"], true=>["_ignored_"]}` , notice false and true are booleans there - and it fails. 

A way to fix this is just to stringify those keys when you make the hash, that should be enough. A test for that would be nice too. :smile:

domcleal commented on test/controllers/api/v1/smart_proxies_controller_test.rb:
```diff
@@ -223,4 +223,12 @@ def setup_import_classes
     end
   end
+
+  test 'should render templates according to api version 2' do
```
version 1

domcleal commented on app/views/api/v2/import_puppetclasses/show.json.rabl:
```diff
@@ -25,3 +26,7 @@ end
   environment.name
 end
+
+node(:ignored_environment, :if => ->(environment) { @changed['ignored'][environment.name].present? && @changed['ignored'][environment.name].match(/_ignored_/) }) do |environment|
```
Also needs a corresponding `ignored_puppetclasses` node.

dLobatog replied:
https://github.com/theforeman/foreman/pull/4012/files#diff-fda05e73ec11be4e3f4c60d62f9bd9e7R26 fixed here, GitHub isn't smart enough to notice yet

domcleal commented on app/controllers/concerns/foreman/controller/environments.rb:
```diff
@@ -21,7 +21,13 @@ def import_environments
+      notice_message = _("No changes to your environments detected")
+
+      if @changed['ignored'].present?
+        notice_message << "\n" + _("Ignored environments: %s") % @changed['ignored'].keys.to_sentence
```
The message here is including names of environments that contain _some_ ignored classes (via the filter), not only environments that are entirely ignored, which is misleading.

dLobatog commented:
Seems to still struggle with boolean-like environments:
`
] [I] Completed 500 Internal Server Error in 2500ms (ActiveRecord: 170.2ms)
09:11:39 rails.1   | 2016-11-18T09:11:39 a75456b0 [app] [F]
09:11:39 rails.1   |  | ArgumentError - comparison of FalseClass with true failed:
09:11:39 rails.1   |  |   app/views/common/_puppetclasses_or_envs_changed.html.erb:32:in`sort'
`

(the ones I have are just called 'yes' and 'no')

bastilian commented:
I was looking into the issue yesterday as well. Stringifying keys would prevent a 500, but The problem would be that they already come as `true` and `false`, so they end up as `"true"`, and would make it impossible to ignore `yes` or 'no' envs.

I'm not sure if it would be a good idea to "escape" 'yes' and 'no' (and other truthy/falsy), before handing it to YAML (just at this instance). Doing it after will not work, since all that were any of the truthy values will then result in just "true".

What do you think @dLobatog?

dLobatog commented:
@bastilian I'd say in that case, detect when the key is a boolean and show that as an error to the user, either in the list of ignored environments or in a 'notice' popup. 

Another solution could be to purposefully make Foreman look for those kind of keys in `ignored_environments.yml` and warn you about the specific env, but it's probably not worth the trouble (I don't expect that many people to have these envs).

bastilian commented:
Showing a notice describing the circumstances is I think a better approach.

I'll add a notice when booleans are found, pointing out that keys were stringified and results in them probably not being found. Plus advice them to quote these names in the yml.

Does that sound like a plan, @dLobatog?

dLobatog commented:
That sounds good to me - no need to get too much into detail,  it's fine to say that boolean environments are not imported properly and point to the UI. Thanks @bastilian!

bastilian commented:
All change requests should now be addressed. The warning in case Boolean values as environment names are found is now shown as well when importing:

![screen shot 2017-01-03 at 09 29 42](https://cloud.githubusercontent.com/assets/7757/21602751/b69d43b2-d197-11e6-8cff-8c4b0b3d3c96.png)

The warning might be a bit long and for some users it might be very vague. Would it maybe be better to add a proper note to the "[Classes](https://theforeman.org/manuals/1.13/index.html#4.2.2Classes)" section in the manual, shorten the warning and link to the note?

bastilian committed:
Fixes #6502 - List ignored classes and environments on import

When importing environments and classes from puppet, ignored
environments and classes defined in ignored_environments.yml
aren't aparent.

This will add them to the list to import as ignored and
should address issues like #6502.

dLobatog commented:
@domcleal Anything missing here? It seems to work well

dLobatog commented:
Thanks @bastilian @domcleal  - let me know if you were missing anything, I checked your last comments and they seemed to be addressed, didn't want this to PR get stuck
